File: JavaFundamentals/ExamPractice/src/org/softuni/TheHeiganDance.java
package org.softuni;

// Java Fundamentals Retake - 26 October 2015

import java.awt.Point;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

public class TheHeiganDance {

    static class Spell {
        String name;
        int longevity;
        int damagePerTurn;
        Point position;

        public Spell(String name, Point position) {
            this.name = name;
            this.position = position;

            if (name.equals("Cloud")) {
                longevity = 2;
                damagePerTurn = 3500;
            } else {
                longevity = 1;
                damagePerTurn = 6000;
            }
        }
    }

    public static void main(String[] args) {

        Scanner in = new Scanner(System.in);
        double heiganHealth = 3000000;
        int playerHealth = 18500;
        Point playerPosition = new Point(7, 7);
        double playerDamage = Double.parseDouble(in.nextLine());
        List<Spell> activeSpells = new ArrayList<>();
        boolean isPlayerDead = false;
        boolean isHeighanDead = false;
        String killerSpell = "";

        while (true) {
            heiganHealth -= playerDamage;
            if (heiganHealth <= 0) {
                isHeighanDead = true;
            }

            String[] params = in.nextLine().split("\\s+");
            String spellName = params[0];
            Point damagePosition = new Point(Integer.parseInt(params[1]), Integer.parseInt(params[2]));

            if (!isHeighanDead) {
                if(isInDamageArea(damagePosition, playerPosition)) {
                    playerPosition = movePlayer(playerPosition, "up", damagePosition);
                    if (isInDamageArea(damagePosition, playerPosition)) {
                        playerPosition = movePlayer(playerPosition, "right", damagePosition);
                        if (isInDamageArea(damagePosition, playerPosition)) {
                            playerPosition = movePlayer(playerPosition, "down", damagePosition);
                            if (isInDamageArea(damagePosition, playerPosition)) {
                                playerPosition = movePlayer(playerPosition, "left", damagePosition);
                                if (isInDamageArea(damagePosition, playerPosition)) {
                                    activeSpells.add(new Spell(spellName, damagePosition));
                                }
                            }
                        }
                    }
                }
            }

            if (!isPlayerDead) {
                for (Spell spell : activeSpells.stream().collect(Collectors.toList())) {
                    spell.longevity--;
                    int damage = spell.damagePerTurn;
                    playerHealth -= damage;
                    if (!isPlayerDead) {
                        if (playerHealth <= 0) {
                            killerSpell = spell.name;
                            isPlayerDead = true;
                        }
                    }
                    if (spell.longevity <= 0) {
                        activeSpells.remove(spell);
                    }
                }
            }

            if (isHeighanDead || isPlayerDead) {
                break;
            }
        }

        if (isHeighanDead) {
            System.out.println("Heigan: Defeated!");
        } else {
            System.out.printf("Heigan: %.2f\r\n", heiganHealth);
        }
        if (isPlayerDead) {
            System.out.println("Player: Killed by " + (killerSpell.equals("Eruption") ? "Eruption" : "Plague Cloud"));
        } else {
            System.out.println("Player: " + playerHealth);
        }
        System.out.println("Final position: " + playerPosition.x + ", " + playerPosition.y);

    }

    private static Point movePlayer(Point playerPosition, String direction, Point damagePosition) {
        Point newPosition = new Point();
        switch (direction) {
            case "up":
                newPosition = new Point(playerPosition.x - 1, playerPosition.y);
                break;
            case "right":
                newPosition = new Point(playerPosition.x, playerPosition.y + 1);
                break;
            case "down":
                newPosition = new Point(playerPosition.x + 1, playerPosition.y);
                break;
            case "left":
                newPosition = new Point(playerPosition.x, playerPosition.y - 1);
                break;
        }
        if (isAtWall(newPosition) || isInDamageArea(damagePosition, newPosition)) {
            return playerPosition;
        } else {
            return newPosition;
        }
    }

    static boolean isInDamageArea(Point damagePosition, Point playerPosition) {
        return playerPosition.x >= damagePosition.x - 1 && playerPosition.x <= damagePosition.x + 1
                && playerPosition.y >= damagePosition.y - 1 && playerPosition.y <= damagePosition.y + 1;
    }

    static boolean isAtWall (Point player) {
        return player.x < 0 || player.x > 15 || player.y < 0 || player.y > 15;
    }
}

File: C#-WebDeveloper-3.0/Advanced-C#-May-2016/Exercises/SetsAndDictionaries/TheHeiganDance/TheHeiganDance.cs
namespace ProblemsWithMatrices
{
    using System;
    using System.Collections.Generic;
    using System.Drawing;
    using System.Text.RegularExpressions;

    public class Character
    {
        private double hp;

        private Point position;

        public bool IsDead => this.hp <= 0;

        public double Hp => this.hp;

        public double Damage { get; private set; }

        public double DamageForNextTurn { get; set; }

        public Point Position => this.position;

        public string KilledBy { get; set; }

        public Character(double hp, double damage)
        {
            this.KilledBy = null;
            this.hp = hp;
            this.Damage = damage;
        }

        public Character(double hp, double damage, Point position)
            : this(hp, damage)
        {
            this.position = position;
        }

        private void Move(string direction, Point spellPosition)
        {
            Point positionBeforeMove = this.position;

            switch (direction)
            {
                case "up":
                    this.position = new Point(this.position.X - 1, this.position.Y);
                    break;
                case "right":
                    this.position = new Point(this.position.X, this.position.Y + 1);
                    break;
                case "left":
                    this.position = new Point(this.position.X, this.position.Y - 1);
                    break;
                case "down":
                    this.position = new Point(this.position.X + 1, this.position.Y);
                    break;
                default:
                    throw new ArgumentOutOfRangeException();
            }

            if (this.IsInDamageArea(spellPosition) || this.IsAtWall())
            {
                this.position = positionBeforeMove;
            }
        }

        public void AttemptEvasion(string spell, Point spellPosition)
        {
            var directions = new Queue<string>(new[] { "up", "right", "down", "left" });

            while (directions.Count != 0 && (this.IsInDamageArea(spellPosition) || this.IsAtWall()))
            {
                this.Move(directions.Dequeue(), spellPosition);
            }

            if (this.IsInDamageArea(spellPosition))
            {
                switch (spell)
                {
                    case "Eruption":
                        this.TakeDamage(6000, spell);
                        break;
                    case "Cloud":
                        this.TakeDamage(3500, spell);
                        this.DamageForNextTurn = 3500;
                        break;
                    default:
                        throw new ArgumentOutOfRangeException();
                }
            }
        }

        public void TakeDamage(double attackerDamage, string spell)
        {
            this.hp -= attackerDamage;

            if (this.KilledBy == null && this.hp <= 0)
            {
                this.KilledBy = spell;
            }
        }

        private bool IsAtWall()
        {
            return this.position.X < 0 || this.position.X > 14 || this.position.Y < 0 || this.position.Y > 14;
        }

        private bool IsInDamageArea(Point damagePosition)
        {
            return this.position.X >= damagePosition.X - 1 && this.position.X <= damagePosition.X + 1
                   && this.position.Y >= damagePosition.Y - 1 && this.position.Y <= damagePosition.Y + 1;
        }
    }

    class TheHeiganDance
    {
        static void Main(string[] args)
        {
            var playerDamage = double.Parse(Console.ReadLine());
            var player = new Character(18500, playerDamage, new Point(7, 7));
            var heigan = new Character(3000000, 0);

            while (!player.IsDead && !heigan.IsDead)
            {
                var parameters = Regex.Split(Console.ReadLine(), @"\s+");
                var spellName = parameters[0];
                var spellRow = int.Parse(parameters[1]);
                var spellCol = int.Parse(parameters[2]);

                if (player.DamageForNextTurn > 0)
                {
                    player.TakeDamage(player.DamageForNextTurn, "Cloud");
                    player.DamageForNextTurn = 0;
                }

                heigan.TakeDamage(player.Damage, "Player");
                if (heigan.IsDead || player.IsDead)
                {
                    break;
                }

                player.AttemptEvasion(spellName, new Point(spellRow, spellCol));
            }

            Console.WriteLine(heigan.IsDead ? "Heigan: Defeated!" : $"Heigan: {heigan.Hp:f2}");

            Console.WriteLine(
                player.IsDead
                    ? $"Player: Killed by {(player.KilledBy.Equals("Cloud") ? "Plague Cloud" : "Eruption")}"
                    : $"Player: {player.Hp}");

            Console.WriteLine($"Final position: {player.Position.X}, {player.Position.Y}");
        }
    }
}